### PR TITLE
Try to use replaceText instead of typeText if the latter fails

### PR DESCRIPTION
**Pull request by mykola-mokhnach**
Addresses https://github.com/appium/ruby_lib/issues/598#issuecomment-392232221

### CODE REVIEW

**Comment by KazuCocoa** [comment]
Could input texts 👍 but it was broken and you can see `P��k��m��n` in the page_source. 
Expected is `Pökémön` as same as uiautomator2. I set `unicodeKeyboard: true, resetKeyboard: true,` [link](https://github.com/appium/appium-espresso-driver/blob/234d2948925a51aacbb097dffbaf014f7b4c3bf4/lib/driver.js#L333) but something happens when handle unicode, it looks. Do we need like https://github.com/appium/appium-uiautomator2-server/blob/28be549849da6ebbfd61791ff594b608958fa4b4/app/src/main/java/io/appium/uiautomator2/utils/ElementHelpers.java#L175 ?

- logcat
```
[HTTP] <-- POST /wd/hub/session/24a6b78c-3771-4986-8115-f57b1b66d42f/element 200 58 ms - 166
[HTTP]
[HTTP] --> POST /wd/hub/session/24a6b78c-3771-4986-8115-f57b1b66d42f/element/8240f1ee-a33d-4386-be42-8386fda58b17/value
[HTTP] {"value":["P","ö","k","é","m","ö","n"],"text":"Pökémön"}
[W3C] Driver proxy active, passing request on via HTTP proxy
[debug] [JSONWP Proxy] Proxying [POST /wd/hub/session/24a6b78c-3771-4986-8115-f57b1b66d42f/element/8240f1ee-a33d-4386-be42-8386fda58b17/value] to [POST http://localhost:8080/session/b905560c-01a4-4948-93f5-97aac085cd13/element/8240f1ee-a33d-4386-be42-8386fda58b17/value] with body: {"value":["P","ö","k","é","m","ö","n"],"text":"Pökémön"}
[debug] [JSONWP Proxy] Got response with status 200: {"id":"f311f449-869c-4049-a7bf-105a4d04994b","sessionId":"b905560c-01a4-4948-93f5-97aac085cd13","status":0,"value":null}
[JSONWP Proxy] Replacing sessionId b905560c-01a4-4948-93f5-97aac085cd13 with 24a6b78c-3771-4986-8115-f57b1b66d42f
[HTTP] <-- POST /wd/hub/session/24a6b78c-3771-4986-8115-f57b1b66d42f/element/8240f1ee-a33d-4386-be42-8386fda58b17/value 200 505 ms - 120
[HTTP]
[HTTP] --> POST /wd/hub/session/24a6b78c-3771-4986-8115-f57b1b66d42f/element
[HTTP] {"using":"id","value":"io.appium.android.apis:id/left_text_edit"}
[W3C] Driver proxy active, passing request on via HTTP proxy
[debug] [JSONWP Proxy] Proxying [POST /wd/hub/session/24a6b78c-3771-4986-8115-f57b1b66d42f/element] to [POST http://localhost:8080/session/b905560c-01a4-4948-93f5-97aac085cd13/element] with body: {"using":"id","value":"io.appium.android.apis:id/left_text_edit"}
[debug] [JSONWP Proxy] Got response with status 200: {"id":"88bf0f57-5d31-4aa8-840b-e2b3813f5bb2","sessionId":"b905560c-01a4-4948-93f5-97aac085cd13","status":0,"value":{"ELEMENT":"dfbb4a9a-b771-4335-adf9-2300f913e5a7"}}
[JSONWP Proxy] Replacing sessionId b905560c-01a4-4948-93f5-97aac085cd13 with 24a6b78c-3771-4986-8115-f57b1b66d42f
[HTTP] <-- POST /wd/hub/session/24a6b78c-3771-4986-8115-f57b1b66d42f/element 200 79 ms - 166
[HTTP]
```

- page_source
```
<?xml version="1.0" encoding="UTF-8"?><com.android.internal.policy.DecorView index="0" package="io.appium.android.apis" class="com.android.internal.policy.DecorView" checkable="false" checked="false" clickable="false" enabled="true" focusable="false" focused="false" scrollable="false" long-clickable="false" password="false" selected="false" bounds="[0,0][1080,1794]" resource-id="">
  <android.widget.LinearLayout index="0" package="io.appium.android.apis" class="android.widget.LinearLayout" checkable="false" checked="false" clickable="false" enabled="true" focusable="false" focused="false" scrollable="false" long-clickable="false" password="false" selected="false" bounds="[0,0][1080,1794]" resource-id="">
    <android.view.ViewStub index="0" package="io.appium.android.apis" class="android.view.ViewStub" checkable="false" checked="false" clickable="false" enabled="true" focusable="false" focused="false" scrollable="false" long-clickable="false" password="false" selected="false" bounds="[0,0][0,0]" resource-id="android:id/action_mode_bar_stub"/>
    <android.widget.FrameLayout index="1" package="io.appium.android.apis" class="android.widget.FrameLayout" checkable="false" checked="false" clickable="false" enabled="true" focusable="false" focused="false" scrollable="false" long-clickable="false" password="false" selected="false" bounds="[0,63][1080,129]" resource-id="android:id/title_container">
      <android.widget.RelativeLayout index="0" package="io.appium.android.apis" class="android.widget.RelativeLayout" checkable="false" checked="false" clickable="false" enabled="true" focusable="false" focused="false" scrollable="false" long-clickable="false" password="false" selected="false" bounds="[11,64][1069,125]" resource-id="app:id/screen">
        <android.widget.TextView index="0" package="io.appium.android.apis" text="Left is best" class="android.widget.TextView" content-desc="Left is best" checkable="false" checked="false" clickable="false" enabled="true" focusable="false" focused="false" scrollable="false" long-clickable="false" password="false" selected="false" bounds="[11,64][194,115]" resource-id="app:id/left_text"/>
        <android.widget.TextView index="1" package="io.appium.android.apis" text="Right is always right" class="android.widget.TextView" content-desc="Right is always right" checkable="false" checked="false" clickable="false" enabled="true" focusable="false" focused="false" scrollable="false" long-clickable="false" password="false" selected="false" bounds="[740,64][1069,115]" resource-id="app:id/right_text"/>
      </android.widget.RelativeLayout>
    </android.widget.FrameLayout>
    <android.widget.FrameLayout index="2" package="io.appium.android.apis" class="android.widget.FrameLayout" checkable="false" checked="false" clickable="false" enabled="true" focusable="false" focused="false" scrollable="false" long-clickable="false" password="false" selected="false" bounds="[0,129][1080,1794]" resource-id="android:id/content">
      <android.widget.LinearLayout index="0" package="io.appium.android.apis" class="android.widget.LinearLayout" checkable="false" checked="false" clickable="false" enabled="true" focusable="false" focused="false" scrollable="false" long-clickable="false" password="false" selected="false" bounds="[0,129][1080,1794]" resource-id="app:id/screen">
        <android.widget.LinearLayout index="0" package="io.appium.android.apis" class="android.widget.LinearLayout" checkable="false" checked="false" clickable="false" enabled="true" focusable="false" focused="false" scrollable="false" long-clickable="false" password="false" selected="false" bounds="[0,129][1080,349]" resource-id="">
          <android.widget.EditText index="0" package="io.appium.android.apis" text="Left is bestP��k��m��n" class="android.widget.EditText" content-desc="Left is best" checkable="false" checked="false" clickable="true" enabled="true" focusable="true" focused="false" scrollable="false" long-clickable="true" password="false" selected="false" bounds="[0,129][550,349]" resource-id="app:id/left_text_edit"/>
          <android.widget.Button index="1" package="io.appium.android.apis" text="Change Left" class="android.widget.Button" content-desc="Change Left" checkable="false" checked="false" clickable="true" enabled="true" focusable="true" focused="false" scrollable="false" long-clickable="false" password="false" selected="false" bounds="[550,176][807,302]" resource-id="app:id/left_text_button"/>
        </android.widget.LinearLayout>
        <android.widget.LinearLayout index="1" package="io.appium.android.apis" class="android.widget.LinearLayout" checkable="false" checked="false" clickable="false" enabled="true" focusable="false" focused="false" scrollable="false" long-clickable="false" password="false" selected="false" bounds="[0,349][1080,475]" resource-id="">
          <android.widget.EditText index="0" package="io.appium.android.apis" text="Right is always right" class="android.widget.EditText" content-desc="Right is always right" checkable="false" checked="false" clickable="true" enabled="true" focusable="true" focused="false" scrollable="false" long-clickable="true" password="false" selected="false" bounds="[0,349][550,475]" resource-id="app:id/right_text_edit"/>
          <android.widget.Button index="1" package="io.appium.android.apis" text="Change Right" class="android.widget.Button" content-desc="Change Right" checkable="false" checked="false" clickable="true" enabled="true" focusable="true" focused="false" scrollable="false" long-clickable="false" password="false" selected="false" bounds="[550,349][827,475]" resource-id="app:id/right_text_button"/>
        </android.widget.LinearLayout>
      </android.widget.LinearLayout>
    </android.widget.FrameLayout>
  </android.widget.LinearLayout>
</com.android.internal.policy.DecorView>
```

**Comment by KazuCocoa** [comment]
I compared bytecode with ASCII. Then the value we can get in https://github.com/appium/appium-espresso-driver/pull/130/files#diff-eee13ba43e532e8b97653b91b5dcb814R45 has `63` twice.

```
private static final Charset M_UTF7 = Charset.forName("US-ASCII");
...

Logger.debug("====encoded:" + Arrays.toString(params.getValue().getBytes(M_UTF7)));
// => [80, 63, 63, 107, 63, 63, 109, 63, 63, 110]
Logger.debug("====Pökémön:" + Arrays.toString("Pökémön".getBytes(M_UTF7)));
// => [80, 63, 107, 63, 109, 63, 110]
```

**Comment by mykola-mokhnach** [comment]
@KazuCocoa Can you please try now. Also, the logcat output would be handy, since I've added more logging there.

**Comment by KazuCocoa** [comment]
with 1058d16 . thanks for the helpful logs. @mykola-mokhnach 

- Appium server
```
[HTTP] --> POST /wd/hub/session/5ecceedb-0703-4b42-b8d0-480492b9edb5/element/0a825bb4-651d-4854-8546-9829f3057f27/value
[HTTP] {"value":["P","ö","k","é","m","ö","n"],"text":"Pökémön"}
[W3C] Driver proxy active, passing request on via HTTP proxy
[debug] [JSONWP Proxy] Proxying [POST /wd/hub/session/5ecceedb-0703-4b42-b8d0-480492b9edb5/element/0a825bb4-651d-4854-8546-9829f3057f27/value] to [POST http://localhost:8080/session/3007f3ce-2d63-4daf-a571-3e5d3c0ac0ea/element/0a825bb4-651d-4854-8546-9829f3057f27/value] with body: {"value":["P","ö","k","é","m","ö","n"],"text":"Pökémön"}
[debug] [JSONWP Proxy] Got response with status 200: {"id":"ef1d6eda-c4e9-46e5-9dea-df3821ada8b9","sessionId":"3007f3ce-2d63-4daf-a571-3e5d3c0ac0ea","status":0,"value":null}
[JSONWP Proxy] Replacing sessionId 3007f3ce-2d63-4daf-a571-3e5d3c0ac0ea with 5ecceedb-0703-4b42-b8d0-480492b9edb5
[HTTP] <-- POST /wd/hub/session/5ecceedb-0703-4b42-b8d0-480492b9edb5/element/0a825bb4-651d-4854-8546-9829f3057f27/value 200 412 ms - 120
[HTTP]
```

- logcat
```
05-27 02:44:14.359 1371-1371/? W/audio_hw_generic: Not supplying enough data to HAL, expected position 2212151 , only wrote 2175840
05-27 02:44:18.071 6203-6257/io.appium.android.apis D/appium: Got raw post data: {"value":["P","��","k","��","m","��","n"],"text":"P��k��m��n"}
05-27 02:44:18.075 6203-6203/io.appium.android.apis I/ViewInteraction: Checking 'MatchesViewAssertion{viewMatcher=is displayed on the screen to the user}' assertion on view Looked for element with View android.widget.EditText{66204fa VFED..CL. .F...... 0,0-550,126 #7f0b006d app:id/left_text_edit}
05-27 02:44:18.083 6203-6257/io.appium.android.apis I/appium: Retrieving element 0a825bb4-651d-4854-8546-9829f3057f27
05-27 02:44:18.087 6203-6203/io.appium.android.apis I/ViewInteraction: Checking 'MatchesViewAssertion{viewMatcher=is displayed on the screen to the user}' assertion on view Looked for element with View android.widget.EditText{66204fa VFED..CL. .F...... 0,0-550,126 #7f0b006d app:id/left_text_edit}
05-27 02:44:18.095 6203-6203/io.appium.android.apis I/ViewInteraction: Performing 'type text(P��k��m��n)' action on view Looked for element with View android.widget.EditText{66204fa VFED..CL. .F...... 0,0-550,126 #7f0b006d app:id/left_text_edit}
05-27 02:44:18.178 1379-5084/? D/gralloc_ranchu: gralloc_alloc: Creating ashmem region of size 45056
05-27 02:44:18.180 1379-5084/? I/chatty: uid=1000(system) HwBinder:1379_3 identical 1 line
05-27 02:44:18.181 1379-5084/? D/gralloc_ranchu: gralloc_alloc: Creating ashmem region of size 45056
05-27 02:44:18.189 6203-6225/io.appium.android.apis D/EGL_emulation: eglMakeCurrent: 0xaab06080: ver 3 0 (tinfo 0xaab03ba0)
05-27 02:44:18.438 6203-6257/io.appium.android.apis D/appium: Trying replaceText action as a workaround to type the 'P��k��m��n' text into the input field
05-27 02:44:18.440 6203-6203/io.appium.android.apis I/ViewInteraction: Performing 'getting view in application' action on view Looked for element with View android.widget.EditText{66204fa VFED..CL. .F...... 0,0-550,126 #7f0b006d app:id/left_text_edit}
05-27 02:44:18.443 6203-6257/io.appium.android.apis D/appium: Current input field's text: 'Left is best'
05-27 02:44:18.445 6203-6203/io.appium.android.apis I/ViewInteraction: Performing 'replace text' action on view Looked for element with View android.widget.EditText{66204fa VFED..CL. .F...... 0,0-550,126 #7f0b006d app:id/left_text_edit}
05-27 02:44:18.447 6227-6227/io.appium.android.ime I/AppiumUnicodeIME: onStartInput
05-27 02:44:18.453 6203-6225/io.appium.android.apis D/EGL_emulation: eglMakeCurrent: 0xaab06080: ver 3 0 (tinfo 0xaab03ba0)
05-27 02:44:18.454 1564-1580/system_process E/memtrack: Couldn't load memtrack module
05-27 02:44:18.454 1564-1580/system_process W/android.os.Debug: failed to get memory consumption info: -1
05-27 02:44:18.458 6203-6257/io.appium.android.apis D/appium: Finished processing POST request for '/session/3007f3ce-2d63-4daf-a571-3e5d3c0ac0ea/element/0a825bb4-651d-4854-8546-9829f3057f27/value'
05-27 02:44:18.467 6203-6225/io.appium.android.apis D/EGL_emulation: eglMakeCurrent: 0xaab06080: ver 3 0 (tinfo 0xaab03ba0)
```

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/handlers/SendKeys.java [hunk 1]
```diff
@@ -65,6 +68,22 @@ public Void handle(TextParams params) throws AppiumException {
             viewInteraction.perform(typeText(value));
         } catch (PerformException e) {
             throw new InvalidElementStateException("sendKeys", params.getElementId(), e);
+        } catch (RuntimeException e) {
+            if (!e.getMessage().contains("IME does not understand how to translate")) {
+                throw e;
```
Comment (jlipps) [inline espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/handlers/SendKeys.java]: the way i am reading this control flow, the only way we will try replaceText is if we get a `RuntimeException`  with the content "IME does not understand how to translate". Is that what's intended? Or did we want to try replaceText in other failure scenarios as well?

Comment (mykola-mokhnach) [reply espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/handlers/SendKeys.java]: yes, we only want to use replaceText for this particular case

**Comment by mykola-mokhnach** [comment]
@dpgraham Can you please publish base driver, so we can if unicode chars in the body are properly parsed by NanoHTTP?

**Comment by mykola-mokhnach** [comment]
@KazuCocoa Ot you can probably add the patch https://github.com/appium/appium-base-driver/pull/214 to your local codebase just for now to verify that the fix works as expected. Thanks.

**Comment by KazuCocoa** [comment]
I applied the base-driver in my local and run this branch. It worked well 👍 
image
Thanks! @mykola-mokhnach

**Pull request merged**